README.md
# back-to/plugins

[![TravisCI build status][travisci-build-status-badge]][travisci-build-status]

Plugins in this Repository are built for **Streamlink 0.14.2+20.gc394b41**
don't expect them to work with the **stable version** of Streamlink.

Submit Questions and Plugin Issue [here][github-issues].

A lot of website are loading to many unnecessary scripts
which slows the computer for no good reason.

The **main purpose** of these plugins is the ability to watch the Video
in a **Videoplayer** such as VLC or MPV instead of a Webbrower.

### GIT source

- [Github][github]
- [Gitlab][gitlab] (mirror)

# Plugin Matrix

`resolve.py` is the main plugin of this repo, it is a **generic Streamlink Plugin**
which will **try** to find a valid Video URL on every website.

You can get a **guide** of every plugin command with `streamlink -h`.

For the other plugins, just go throw them and see if you need something.

# Guide

### How can I use the Streamlink Development version?

- [Linux][install-linux]
- [Windows][install-windows]

### How can I download these Plugins?

If you are on Linux, the best way is to use `git`

If you are on something else you can just download the [ZIP file][repo-zip]

### How can I use these Plugins?

The best way is to sideload them with Streamlink,
you will have to create a new folder and put them in it.

| Platform          | Location           |
| -------------     | -------------      |
| Unix-like (POSIX) | $XDG_CONFIG_HOME/streamlink/plugins |
| Windows           | %APPDATA%\streamlink\plugins        |

For more details see [here][streamlink-sideloading]

### How can I use these Plugins with Kodi?

[repository.back-to][repository.back-to] must be installed

- open in Kodi **Add-ons / Add-on browser**
- click on **Install from repository**
- click on **back-to repository**
- click on **service**
- click on **LiveProxy**
- click on **Dependencies**
- search for **back-to plugins**
- install it
- done

### Linux Quick Install Guide

you will have to create the default streamlink config folder


```sh
mkdir $HOME/.config/streamlink
```

clone it and make a symbolic link

```sh
git clone https://github.com/back-to/plugins.git
cd plugins
ln -s "$(pwd)/plugins/" "$HOME/.config/streamlink/"
```

you can update the plugins with git in the folder where it was cloned

```sh
git pull origin master
```

you can also put your own custom plugins in the plugins folder,
if they are named `hidden_NAME.py` they won't interfere with the git command.

  [github-issues]: https://github.com/back-to/plugins/issues
  [github]: https://github.com/back-to/plugins
  [gitlab]: https://gitlab.com/back-to/plugins
  [install-linux]: https://streamlink.github.io/latest/install.html#source-code
  [install-windows]: https://bintray.com/streamlink/streamlink-nightly/streamlink/_latestVersion/#files
  [repo-zip]: https://github.com/streamlink/streamlink/archive/master.zip
  [repository.back-to]: https://github.com/back-to/repo/raw/master/repository.back-to/repository.back-to-5.0.0.zip
  [streamlink-sideloading]: https://streamlink.github.io/latest/cli.html#sideloading-plugins
  [travisci-build-status-badge]: https://travis-ci.org/back-to/plugins.svg?branch=master
  [travisci-build-status]: https://travis-ci.org/back-to/plugins


CHANGELOG.md
# Changelog

## 2018-08-19
### Changed
- **plugins.fc2**: changed websocket param
- **plugins.showup**: use wss instead of ws
- **plugins.resolve**: iframe regex cleanup
- **plugins.resolve**: changes some log.debug to log.trace
- **plugins.resolve**: Fixed wrong regex for unescape_type iframes
- **plugins.resolve**: playlist_re update for \ before the last "

### Removed
- **plugins.resolve**: removed p.a.c.k.e.r support.
- plugin test files

## 2018-08-14
### Changed
- **plugins.showup**: ws host update

## 2018-08-07
### Changed
- **plugins.resolve**: removed itertags, use re

## 2018-07-12
### Added
- **plugins.rutube**: metadata for the --title command

### Changed
- All Plugins: use self.session.http instead of http

## 2018-07-10
### Added
- README: New Guide
- plugins.myfreecams: support for wzobs servers
- plugins.myfreecams: DASH streams but only for h5video servers
                      and --myfreecams-dash must be used.

### Changed
- plugins.myfreecams: some log error's are now PluginError's,
                      the console output is a bit different,
                      also added more debug messages.

## 2018-07-09
### Added
- **common_packer**: Unpacker for Dean Edward's p.a.c.k.e.r
  The p.a.c.k.e.d feature is experimental and might not work for every packed m3u8 URL.
  It is used with the resolve plugin, but can also be used for other plugins.
- **plugins.resolve**: User input for iframes
  If a website has more than one iframe,
  the user will be asked to choose which iframe should be used.
- **plugins.resolve**: support for unescaped m3u8 URLs

### Changed
- **plugins.resolve**: Added a chat domain for the static blocklist

## 2018-07-07
### Added
- **plugins.resolve**: basic support for plugin.get_title()

### Changed
- **plugins.resolve**: moved self.url into init
- **plugins.resolve**: use res_text as self.html_text

## 2018-07-06
### Added
- **plugins.otr**: New plugin for ОТР

### Changed
- **Streamlink 0.14.2+20.gc394b41** is now required for hlssession.py, resolve.py
- **plugins.resolve**: speedup compare_url_path

## 2018-07-04
### Added
- **plugins.resolve**: New command --resolve-blacklist-filepath

### Changed
- **plugins.resolve**: Block the path end '/novideo.mp4'
- **README** - Linux Guide update
- **README** - Plugin Guide cleanup, `streamlink -h` should be used

## 2018-07-01
### Added
- debug log message for the plugin version


plugins/fc2.py
# -*- coding: utf-8 -*-
import json
import logging
import re

from datetime import datetime
from requests.utils import dict_from_cookiejar
from threading import Thread, Timer
from websocket import create_connection

from streamlink.exceptions import PluginError
from streamlink.plugin import Plugin, PluginArgument, PluginArguments
from streamlink.plugin.api import useragents, validate
from streamlink.stream import RTMPStream

log = logging.getLogger(__name__)


class FC2(Plugin):

    url_login = 'https://secure.id.fc2.com/?mode=login&switch_language=en'
    url_member_api = 'https://live.fc2.com/api/memberApi.php'
    url_server = 'https://live.fc2.com/api/getControlServer.php'

    _url_re = re.compile(r'''https?://live\.fc2\.com/(?P<user_id>\d+)/?$''')

    _version_schema = validate.Schema({
        'status': int,
        'data': {
            'channel_data': {
                'channelid': validate.text,
                'userid': validate.text,
                'adult': int,
                'login_only': int,
                'version': validate.text,
                'fee': int,
            },
            'user_data': {
                'is_login': int,
                'userid': int,
                'fc2id': int,
                'name': validate.text,
                'point': int,
                'adult_access': int,
                'recauth': int,
            }
        }
    })

    count = 0
    count_ping = 0

    host_data = ''
    host_found = False

    arguments = PluginArguments(
        PluginArgument(
            'username',
            requires=['password'],
            metavar='USERNAME',
            help='The username associated with your FC2 account.',
        ),
        PluginArgument(
            'password',
            sensitive=True,
            metavar='PASSWORD',
            help='A FC2 account password to use with --fc2-username',
            prompt='Enter FC2 password'
        ),
        PluginArgument(
            'purge-credentials',
            action='store_true',
            help='''
        Purge cached FC2 credentials to initiate a new session
        and reauthenticate.
        '''
        )
    )

    @classmethod
    def can_handle_url(cls, url):
        return cls._url_re.match(url)

    def _login(self, username, password):
        '''login and update cached cookies'''
        log.debug('login ...')
        self.session.http.get(self.url)
        data = {
            'pass': password,
            'email': username,
            'done': 'livechat',
            'keep_login': 1
        }

        self.session.http.post(self.url_login, data=data, allow_redirects=True)
        cookies_list = self.save_cookies()

        return self.cmp_cookies_list(cookies_list)

    def _get_version(self, user_id):
        data = {
            'user': 1,
            'channel': 1,
            'profile': 1,
            'streamid': int(user_id)
        }
        res = self.session.http.post(self.url_member_api, data=data)
        res_data = self.session.http.json(res, schema=self._version_schema)
        channel_data = res_data['data']['channel_data']
        user_data = res_data['data']['user_data']

        if (channel_data['login_only'] != 0 and user_data['is_login'] != 1):
            raise PluginError('A login is required for this stream.')

        if channel_data['fee'] != 0:
            raise PluginError('Only streams without a fee are supported.')

        version = channel_data['version']
        if user_data['is_login']:
            log.info('Logged in as {0}'.format(user_data['name']))
        log.debug('Found version: {0}'.format(version))
        return version

    def payload_msg(self, name):
        ''' Format the WebSocket message '''
        self.count_ping += 1
        payload = json.dumps(
            {
                'name': str(name),
                'arguments': {},
                'id': int(self.count_ping)
            }
        )
        return payload

    def _get_ws_url(self, user_id, version):
        log.debug('_get_ws_url ...')
        data = {
            'channel_id': user_id,
            'channel_version': version,
            'client_type': 'pc',
            'client_app': 'browser'
        }

        res = self.session.http.post(self.url_server, data=data)
        w_data = self.session.http.json(res)
        if w_data['status'] == 11:
            raise PluginError('The broadcaster is currently not available')

        ws_url = '{0}?control_token={1}&comment=0'.format(
            w_data['url'], w_data['control_token'])
        log.debug('WS URL: {0}'.format(ws_url))
        return ws_url

    def _get_ws_data(self, ws_url):
        log.debug('_get_ws_data ...')
        ws = create_connection(ws_url)
        ws.send(self.payload_msg('get_media_server_information'))

        def ws_ping():
            ''' ping the WebSocket '''
            if ws.connected is True:
                t1 = Timer(30.0, ws_ping)
                t1.daemon = True
                t1.start()
                ws.send(self.payload_msg('heartbeat'))

        def ws_recv():
            ''' print WebSocket messages '''
            while True:
                self.count += 1
                data = json.loads(ws.recv())
                time_utc = datetime.utcnow().strftime('%H:%M:%S UTC')
                if data['name'] not in ['comment', 'ng_commentq',
                                        'user_count', 'ng_comment']:
                    log.debug('{0} - {1} - {2}'.format(
                        time_utc, self.count, data['name']))

                if (data['name'] == '_response_'
                        and data['arguments'].get('host')):
                    log.debug('Found host data')
                    self.host_data = data
                    self.host_found = True
                elif data['name'] == 'media_connection':
                    log.debug('successfully opened stream')
                elif data['name'] == 'control_disconnection':
                    if self.count <= 30:
                        # User with points restricted program being broadcasted
                        self.count = 30
                    if data.get('arguments').get('code') == 4512:
                        log.debug('Disconnected from Server')
                    break
                elif data['name'] == 'publish_stop':
                    log.debug('Stream ended')
                elif data['name'] == 'channel_information':
                    if data['arguments'].get('fee') != 0:
                        log.error('Stream requires a fee now.'.format(
                            data['arguments'].get('fee')))
                        break
                elif data['name'] == 'media_disconnection':
                    if data.get('arguments').get('code') == 104:
                        log.warning('Disconnected. '
                                    'Multiple connections has been detected.')
                    elif data.get('arguments').get('code'):
                        log.debug('error code {0}'.format(
                            data['arguments']['code']))

            ws.close()

        # WebSocket background process
        ws_ping()
        t2 = Thread(target=ws_recv)
        t2.daemon = True
        t2.start()

        # wait for the WebSocket
        host_timeout = False
        while self.host_found is False:
            if self.host_found is True:
                break
            if self.count >= 30:
                host_timeout = True
                break

        log.debug('host_timeout is {0}'.format(host_timeout))
        if host_timeout:
            return False
        return True

    def _get_rtmp(self, data):
        log.debug('_get_rtmp ...')

        app = '{0}?media_token={1}'.format(
            data['application'], data['media_token'])
        host = data['host']

        params = {
            'app': app,
            'flashVer': 'WIN 29,0,0,140',
            'swfUrl': 'https://live.fc2.com/swf/liveVideo.swf',
            'tcUrl': 'rtmp://{0}/{1}'.format(host, app),
            'live': 'yes',
            'pageUrl': self.url,
            'playpath': data['play_rtmp_stream'],
            'host': host,
        }
        yield 'live', RTMPStream(self.session, params)

    def cmp_cookies_list(self, cookies_list):
        required_cookies = [
            'FCSID', 'fcu', 'fgcv', 'glgd_val',
            'login_status', 'PHPSESSID', 'secure_check_fc2',
        ]
        count = 0
        for c in required_cookies:
            if c in cookies_list:
                count += 1
        log.debug('Same Cookies: {0}'.format(count))
        return (count == len(required_cookies))

    def _get_streams(self):
        log.debug('Version 2018-08-19')
        log.info('This is a custom plugin.')

        if self.options.get('purge_credentials'):
            self.clear_cookies()
            log.info('All credentials were successfully removed.')

        self.session.http.headers.update({
            'User-Agent': useragents.FIREFOX,
            'Referer': self.url
        })

        cookies_list = []
        for k in dict_from_cookiejar(self.session.http.cookies):
            cookies_list.append(k)
        _authed = self.cmp_cookies_list(cookies_list)

        login_username = self.get_option('username')
        login_password = self.get_option('password')

        if _authed:
            log.info('Attempting to authenticate using cached cookies')
        elif (not _authed and login_username and login_password):
            if not self._login(login_username, login_password):
                log.error('Failed to login, check your username/password')

        match = self._url_re.match(self.url)
        if not match:
            return

        user_id = match.group('user_id')

        version = self._get_version(user_id)
        ws_url = self._get_ws_url(user_id, version)
        if self._get_ws_data(ws_url):
            return self._get_rtmp(self.host_data['arguments'])


__plugin__ = FC2
